vendor/net_gearman/Net/Gearman/AtomWorker.php

```diff
@@ -0,0 +1,152 @@
+<?php
+
+/*
+ * This file is part of the Access to Memory (AtoM) software.
+ *
+ * Access to Memory (AtoM) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Access to Memory (AtoM) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+class AtomNetGearmanWorker extends Net_Gearman_Worker
+{
+    protected $jobsCompleted = 0;
+    protected $maxJobCount = 0;
+
+    public function setMaxJobCount(int $maxJobCount)
+    {
+        if ($maxJobCount > 0) {
+            $this->maxJobCount = $maxJobCount;
+        }
+    }
+
+    public function getJobsCompleted()
+    {
+        return $this->jobsCompleted;
+    }
+
+    /**
+     * Begin working.
+     *
+     * Overrides net_gearman's Worker.php beginWork method. Adds check
+     * $this->maxJobCountReached() and exits work loop if true.
+     *
+     * @param callable $monitor Function to monitor work
+     *
+     * @see Net_Gearman_Connection::send(), Net_Gearman_Connection::connect()
+     * @see Net_Gearman_Worker::doWork(), Net_Gearman_Worker::addAbility()
+     */
+    public function beginWork($monitor = null)
```

Comment:
Rather than duplicating the `Net_Gearman_Worker::beginWork()` code, what about using the `$monitor` callback to stop the `$working` loop and kill the worker?  It would require integrating the MySQL connection keep-alive anonymous function from https://github.com/artefactual/atom/blob/077b17da608339dceedaefc68b95bc8f59e50d2c/lib/task/jobs/jobWorkerTask.class.php#L123 but I think it's still worthwhile.

Reply:
This would still require overriding beginWork() as currently the monitor callback only has 2 vars available to it: an idle flag and the last job time available to it. The vendor Worker.php file would still need to be modified here to pass in `$this` in addition to the other two vars:

https://github.com/artefactual/atom/blob/qa/2.x/vendor/net_gearman/Net/Gearman/Worker.php#L298

....then in the callback I could call `maxJobCountReached()` and return false exiting the loop.

I think your suggestion makes a lot of sense, but it will still require overriding beginWork() I think.

Reply:
My understanding, based on https://www.php.net/manual/en/functions.anonymous.php#example-162 (Example 5: Automatic binding of $this), is `$this` is automatically available within closure is defined within the class context.  I think `function () use ($var1, $var2)` can also be used to pass extra variables into the closure scope.

Reply:
And I don't object to overriding `Net_Gearman_Worker::beginWork()`, my objection is with copying and pasting all the code from `Net_Gearman_Worker::beginWork()` into `AtomWorker::beginWork()`. :)

Reply:
This is extremely useful feedback - thanks @djjuhasz.  Expect another commit soon...  😄